유저 관련 API 및 JWT 모듈 구현

seokju2ng opened the pull request:
### 추가된 의존성
- `jsonwebtoken`

### 변경 사항
- User Model
  - 프로필 이미지 url 저장을 위한 컬럼 `imageUrl` 추가
  - 그에 따른 User 시더 데이터 추가

### 추가 사항
- Users API
```http
# 유저 전체 조회
GET /api/users

# 로그인 
POST /api/users/login

# 회원가입
POST /api/users/join

# 토큰 인증
GET /api/users/auth
```

- Error Handling
  - 에러 메시지 추가
    - 로그인 실패
    - 회원가입 실패
    - 유효하지 않은 토큰 검증
  - HTTP status code 추가
    - 401 Unauthorized
    - 403 Forbidden
- jwt Module
  - `sign(user)`: 인증된 회원에 대한 JWT 발급
  - `authentication(token)`: token이 유효한지 검증

seokju2ng committed:
[server] feat: 유저 모델에 프로필 이미지 컬럼 추가

- 프로필 이미지 url 저장을 위한 컬럼 `imageUrl` 추가
- 그에 따른 시더 데이터 추가

seokju2ng committed:
[server] feat: 로그인 모듈 & JWT 발급 구현

- `jsonwebtoken` 모듈 의존
- `userService`의 로그인 모듈 구현
- `jwt` 모듈의 JWT 발급 구현
- 403 Forbidden HTTP status code 추가
- 로그인 실패 에러 핸들링

seokju2ng committed:
[server] feat: 회원가입 모듈, 로그인 & 회원가입 API 구현

- `userService`에 회원가입 모듈 추가
- `/api/users/login` 로그인 API 라우터 추가
- `/api/users/join` 회원가입 API 라우터 추가
- `userController` 추가
- 회원가입 실패(ID가 이미 존재하는 경우) 에러 핸들링 추가

Close #239

seokju2ng committed:
[server] refactor: 로그인 모듈과 JWT 발급 모듈의 의존성 제거

- 기존: 로그인 모듈에서 JWT를 발급해서 리턴해주었음
- 문제:
  1. 로그인 모듈이 JWT 모듈을 의존함
  2. 로그인 모듈이나 JWT 모듈이 각각 독립적으로 다른 곳에서 사용될
     가능성이 있음
- 해결:
  - 로그인 모듈과 JWT 모듈의 의존을 끊음
  - 컨트롤러에서 로그인 모듈 다녀와서 성공하면 JWT 발급 모듈을 통해
  token 발급

seokju2ng committed:
[server] feat: JWT 인증 모듈 구현

- `jwt.authentication(token)` token이 유효한지 검증하는 함수
- 401 Unauthorized HTTP status code 추가
- 유효하지 않은 토큰에 대한 에러 핸들링 추가
- `/api/users/auth?token` 요청에 대한 라우팅 추가

Close #214

seokju2ng committed:
[server] feat: Users 전체 목록 조회 API 구현

- `/api/users` GET 요청 처리하는 API 구현
- 유저의 패스워드만 빼고 전체 목록을 응답

Close #209

skid901 commented on web/server/src/common/jwt.js:
```diff
@@ -0,0 +1,34 @@
+const secret = process.env.JWT_SECRET;
+
+module.exports = {
+  sign: user =>
+    new Promise((resolve, reject) => {
+      jwt.sign(
+        {
+          num: user.num,
+          id: user.id,
+        },
```
```
sign: ({ num, id }) => 
...
    jwt.sign(
      {
        num,
        id
...
)
```
이 메소드의 인자를 구조분해할당해서 받으면 좋을 것 같아요!!

skid901 commented on web/server/src/controllers/users.js:
```diff
@@ -0,0 +1,32 @@
+
+  join: async (req, res) => {
+    const { id, password, name, imageUrl } = req.body;
+    const user = await userService.join({ id, password, name, imageUrl });
+    delete user.dataValues.password;
```
```
delete user.dataValue.password;
```
이 로직을 `userService.join()` 내부에서 수행하도록 해주시면 좋을 것 같아요!!

skid901 commented on web/server/src/controllers/users.js:
```diff
@@ -0,0 +1,32 @@
+    res.status(CREATED).json({ success: true, user });
+  },
+
+  auth: async (req, res) => {
+    const { token } = req.query;
+    const decoded = await jwt.authentication(token);
+    res.status(OK).json({ success: true, user: decoded });
+  },
```
검증 단계에서 토큰에 기록되었던 데이터에 대한 유효성 검사를 하지 않아도 될까요?

seokju2ng replied:
토큰을 발급 받고, 인증을 요청한 사이에 탈퇴와 같은 처리가 이루어지면 유효성 체크가 필요하지 싶은데,
저희는 탈퇴를 구현하지 않을 예정이라 지금은 고려하지 않아도 될 것 같습니다!

skid901 commented on web/server/src/services/users.js:
```diff
@@ -0,0 +1,31 @@
+const userService = {
+  login: async ({ id, password }) => {
+    const user = await User.findOne({ where: { id } });
+    const verify = (user, pwd) => user.password === pwd;
+
+    if (!user) {
+      throw new Error(LOGIN_FAILED);
+    }
+    if (!verify(user, password)) {
+      throw new Error(LOGIN_FAILED);
+    }
```
1. if 문 안의 조건문을 명시적으로 나타내실 목적으로 사용하시는 것이면 `verifyPassword` 로 함수명을 변경하시는 것은 어떠신가요? 또는 선언된 함수가 한번만 사용되고 있기 때문에 함수로 로직을 묶지 않아도 될 것 같아요.
2. 두 if 문을 하나로 만들어도 되지 않을까요?

seokju2ng replied:
`verify()`함수를 조금 변형해서 if문을 하나로 바꾸었습니다!

ahrimy commented on web/server/src/controllers/users.js:
```diff
@@ -0,0 +1,32 @@
+  },
+
+  auth: async (req, res) => {
+    const { token } = req.query;
```
token은 query보다는 Header에 넣어서 보내는게 좋지 않을까요??

seokju2ng replied:
간단히 테스트를 해보기 위해 일단 query string으로 전달했습니다.
현재 토큰을 발급해줄 때도 그냥 응답 바디에 보내주고 있습니당
그 작업은 나중에 맞춰요 !

ahrimy reviewed (approved):
고생하셨습니다!
OAuth 부분이랑 연동하고 인증하는 middleware 추가하면 될것같네용!

seokju2ng committed:
[server] refactor: 리뷰 반영

- jwt 모듈의 sign함수 파라미터를 구조분해할당
- 컨트롤러가 아닌 서비스에서 비즈니스 로직 수행
- 로그인시 에러처리 if문 두개를 하나로 합침
